crypto/threads_pthread.c

```diff
@@ -42,12 +45,562 @@
+#else
+static pthread_mutex_t atomic_sim_lock = PTHREAD_MUTEX_INITIALIZER;
+
+static inline void* fallback_atomic_load_n(void **p)
+{
+    void *ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *(void **)p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_LOAD_N(p, o) fallback_atomic_load_n((void **)p)
+
+static inline void* fallback_atomic_store_n(void **p, void *v)
+{
+    void *ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *p;
+    *p = v;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_STORE_N(p, v, o) fallback_atomic_store_n((void **)p, (void *)v)
+
+static inline void fallback_atomic_store(void **p, void **v)
+{
+    void *ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *p;
+    *p = *v;
+    v = ret;
+    pthread_mutex_unlock(&atomic_sim_lock);
+}
+
+# define ATOMIC_STORE(p, v, o) fallback_atomic_store((void **)p, (void **)v)
+
+static inline void* fallback_atomic_exchange_n(void **p, void *v)
+{
+    void *ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *p;
+    *p = v;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+#define ATOMIC_EXCHANGE_N(p, v, o) fallback_atomic_exchange_n((void **)p, (void *)v)
+
+static inline uint64_t fallback_atomic_add_fetch(uint64_t *p, uint64_t v)
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p += v;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_ADD_FETCH(p, v, o) fallback_atomic_add_fetch(p, v)
+
+static inline uint64_t fallback_atomic_fetch_add(uint64_t *p, uint64_t v)
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *p;
+    *p += v;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_FETCH_ADD(p, v, o) fallback_atomic_fetch_add(p, v)
+
+static inline uint64_t fallback_atomic_sub_fetch(uint64_t *p, uint64_t v)
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p -= v;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_SUB_FETCH(p, v, o) fallback_atomic_sub_fetch(p, v)
+
+static inline uint64_t fallback_atomic_and_fetch(uint64_t *p, uint64_t m)
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p &= m;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_AND_FETCH(p, v, o) fallback_atomic_and_fetch(p, v)
+
+static inline uint64_t fallback_atomic_or_fetch(uint64_t *p, uint64_t m)
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p |= m;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_OR_FETCH(p, v, o) fallback_atomic_or_fetch(p, v)
+#endif
+
+static CRYPTO_THREAD_LOCAL rcu_thr_key;
+
+/*
+ * users is broken up into 2 parts
+ * bits 0-15 current readers
+ * bit 32-63 - ID
```

Comment:
yes, or at least aruably so.  Bits 16-31 were, several revisions ago used to track the number of writers on a qp.  That has since become no longer necessecary, but I would rather keep them reserved here for other potential uses.  The reader count tracks the number of readers that we have holding the lock, so 2^16 seems like more than enough IMHO.